clash-lib/src/Clash/Core/EqSolver.hs

```diff
@@ -124,6 +136,12 @@ isAbsurdAlt
   :: TyConMap
   -> Alt
   -> Bool
+#if MIN_VERSION_base(4,15,0)
+isAbsurdAlt _tcm (DataPat dc _ _,_)
+  -- unsafeCoerce is not absurd in the way intended by /isAbsurdAlt/
+  | dcUniq dc == getKey unsafeReflDataConKey
+  = False
```

Comment:
Why is this only a thing for 9.0?

Reply:
Because of the new implementation of unsafeCoerce based on pattern-matching on `UnsafeRefl` is a base 4.15 and up thing. Before that, unsafeCoerce would be translated to a cast at the GHC core level.